urllib3/util/hpkp.py

```diff
@@ -0,0 +1,336 @@
+        try:
+            value = directive[1]
+        except IndexError:
+            value = None
+
+        if name == 'pin-sha256':
+            value = unquote_string(value)
+            pin_directives.append(value)
+        elif name == 'max-age':
+            # There may only be one max-age directive. Be careful about
+            # policing this, we don't want to accidentally allow more than one.
+            if max_age:
+                raise HPKPError("Multiple max-age directives in PKP header")
+
+            max_age = int(unquote_string(value))
+            if max_age < 0:
+                raise HPKPError("max-age must be positive")
+        elif name == 'includesubdomains':
+            include_subdomains = True
+        elif name == 'report-uri':
+            if report_uri:
+                raise HPKPError("Multiple report-uri directives in PKP header")
+
+            report_uri = unquote_string(value)
+
+    if not pin_directives:
+        raise HPKPError("No pin directives in PKP header")
+
+    # LUKASA: Is the use of time.time() safe here? Probably not.
+    return KnownPinnedHost(
+        domain, pin_directives, max_age, include_subdomains, report_uri, time.time()
+    )
+
+
+def _validate_pin(connection, host, shortcut=True):
+    """
+    Validates that a TLS connection matches a specific pin.
+
+    If shortcut is ``True``, returns as soon as a matching cert is found: do
+    not use the return value in this case.
+
+    Returns True if the pin contains at least one key *not* used in this
+    certificate chain. Returns False if the pin only contains certificates in
+    the chain.
+    """
+    match = False
+    non_match = False
+
+    # We ideally want to grab the whole cert chain here.
+    certificates = connection.sock.get_peer_cert_chain()
+
+    for binary_certificate in certificates:
+        # For each cert in the chain, get a base64-encoded form of the
+        # SHA256 of the public key and check whether it's in the pin list.
+        if _certificate_in_pins(binary_certificate, host):
+            if shortcut:
+                return True
+
+            match = True
+        else:
+            non_match = True
+
+    if not match:
+        raise HPKPError("Failed to validate trust chain with HPKP!")
+
+    return match and non_match
+
+
+def _certificate_in_pins(der_certificate, host):
+    """
+    For a single DER certificate, check whether the KnownPinnedHost has
+    pinned it.
+    """
+    cert = x509.load_der_x509_certificate(
+        der_certificate, default_backend()
+    )
+    key = cert.public_key()
+    public_key = key.public_bytes(Encoding.PEM,
+                                  PublicFormat.SubjectPublicKeyInfo)
+    public_key_base64 = ''.join(public_key.split("\n")[1:-2])
+    public_key_raw = base64.b64decode(public_key_base64)
+    public_key_sha256 = hashlib.sha256(public_key_raw).digest()
+    public_key_sha256_base64 = base64.b64encode(public_key_sha256)
```

Comment:
Next time cryptography releases they'll add `Encoding.DER`, which will allow us to skip this weird trip through PEM (which requires us to do the crazy split-join-b64decode that we do here).